File: Memento/PHP/EditorHistoryCaretaker.php
<?php

require_once('WorkspaceState.php');

class EditorHistoryCaretaker {
    private array $stateList;

    public function add(WorkspaceState $workspaceState) {
        $this->stateList[] = $workspaceState;
    }

    public function get(int $index) { 
        if (count($this->stateList) < 1) {
            return null;
        }

        if (count($this->stateList) < $index + 1) {
            return $this->stateList[count($this->stateList) - 1];
        } else {
            return $this->stateList[$index];
        }
    }
}

File: Memento/Golang/EditorHistoryCaretaker.go
package main

type EditorHistoryCaretaker struct {
	stateList []*WorkspaceState
}

func (editorHistoryCaretaker *EditorHistoryCaretaker) add(workspaceState *WorkspaceState) {
	editorHistoryCaretaker.stateList = append(editorHistoryCaretaker.stateList, workspaceState)
}

func (editorHistoryCaretaker *EditorHistoryCaretaker) get(index int) *WorkspaceState {
	if len(editorHistoryCaretaker.stateList) < 1 {
		return nil
	}

	if len(editorHistoryCaretaker.stateList) < index + 1 {
		return editorHistoryCaretaker.stateList[len(editorHistoryCaretaker.stateList) - 1]
	} else {
		return editorHistoryCaretaker.stateList[index]
	}
}
